File: minecraft/src/main/java/com/spider/minecraft/security/DuressManager.java
package com.spider.minecraft.security;

import com.spider.minecraft.SpiderMinecraft;
import com.spider.minecraft.SpiderMinecraftMod;
import com.spider.minecraft.crypto.KeyManager;
import com.spider.minecraft.network.SessionManager;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * 胁迫 PIN 管理器 — 对齐 Spider 的 duress PIN 特性。
 *
 * <p>由于 Minecraft 聊天栏无法输入隐藏字符，无法实现"登录时输入胁迫 PIN"。
 * 因此实现"命令式胁迫"：
 * <ul>
 *   <li>用户输入 /spiderminecraft duress <pin></li>
 *   <li>如果 PIN 匹配身份文件中的胁迫 PIN 哈希</li>
 *   <li>立即执行 wipe_all_data()（删除 identity.json 和密钥）</li>
 *   <li>自动向服务端发送 COMPROMISED 加密信令</li>
 * </ul>
 *
 * <p>胁迫 PIN 为 6 位数字，存储 SHA-256 哈希，不存储明文。
 */
public class DuressManager {

    private static final Logger LOGGER = LogManager.getLogger(SpiderMinecraft.MOD_NAME);

    private final KeyManager keyManager;
    private final SessionManager sessionManager;

    public DuressManager(KeyManager keyManager, SessionManager sessionManager) {
        this.keyManager = keyManager;
        this.sessionManager = sessionManager;
    }

    /**
     * 触发胁迫流程。
     *
     * @param pin 用户输入的 PIN
     * @return true 如果是胁迫 PIN（已触发擦除），false 如果不是
     */
    public boolean triggerDuress(String pin) {
        if (pin == null || pin.length() != SpiderMinecraft.DURESS_PIN_LENGTH || !pin.matches("\\d+")) {
            return false;
        }

        if (!keyManager.checkDuressPin(pin)) {
            return false;
        }

        LOGGER.warn("[SpiderMinecraft] DURESS PIN DETECTED — initiating wipe protocol");

        // 1. 先发送 COMPROMISED 信令（在密钥还在内存中时签名）
        if (sessionManager != null && sessionManager.isAuthenticated()) {
            sessionManager.sendCompromised();
            // 给服务器一点时间接收
            try { Thread.sleep(500); } catch (InterruptedException ignored) {}
        }

        // 2. 执行 wipe_all_data()
        keyManager.wipeAllData();

        // 3. 断开连接
        if (sessionManager != null) {
            sessionManager.logout();
        }

        LOGGER.warn("[SpiderMinecraft] Wipe protocol complete — all data erased");
        return true;
    }

    /**
     * 设置胁迫 PIN（需要先解锁身份）。
     */
    public boolean setDuressPin(String pin) {
        if (pin == null || pin.length() != SpiderMinecraft.DURESS_PIN_LENGTH || !pin.matches("\\d+")) {
            return false;
        }
        if (!keyManager.isUnlocked()) {
            return false;
        }
        try {
            keyManager.setDuressPin(pin);
            LOGGER.info("[SpiderMinecraft] Duress PIN set");
            return true;
        } catch (Exception e) {
            LOGGER.error("[SpiderMinecraft] Failed to set duress PIN: {}", e.getMessage());
            return false;
        }
    }

    public boolean hasDuressPin() {
        return keyManager.hasDuressPin();
    }
}

File: android/app/src/main/java/com/spider/android/security/DuressManager.kt
package com.spider.android.security

import android.content.Context
import android.util.Log
import com.spider.android.crypto.KeyManager
import com.spider.android.network.SpiderClient
import com.spider.android.session.SessionManager
import com.spider.android.storage.DatabaseHelper

/**
 * 胁迫 PIN 管理器 — 对齐 Spider 的 duress PIN 特性。
 *
 * 功能：
 * - 检测输入的 PIN 是否为胁迫 PIN
 * - 触发胁迫流程：发送 COMPROMISED + 擦除本地数据
 * - 设置/修改胁迫 PIN
 */
class DuressManager(
    private val context: Context,
    private val keyManager: KeyManager,
    private val spiderClient: SpiderClient,
    private val sessionManager: SessionManager
) {

    private val TAG = "DuressManager"

    var onDuressTriggered: (() -> Unit)? = null

    /**
     * 检查 PIN 是否为胁迫 PIN。如果是，触发胁迫流程。
     * @return true 如果是胁迫 PIN（已触发擦除），false 如果不是
     */
    fun checkAndTrigger(pin: String): Boolean {
        if (!keyManager.hasDuressPin()) return false
        if (!keyManager.checkDuressPin(pin)) return false

        Log.w(TAG, "DURESS PIN DETECTED — initiating wipe protocol")
        triggerDuress()
        return true
    }

    /**
     * 触发胁迫流程。
     */
    fun triggerDuress() {
        // 发送 COMPROMISED 信令
        val uuid = keyManager.getUuid()
        val identity = keyManager.getIdentity()
        if (uuid.isNotEmpty() && identity != null && spiderClient.isConnected) {
            val signData = "compromised|$uuid|${System.currentTimeMillis() / 1000}"
            // 注意：这里需要 cryptoManager，但为了避免循环依赖，通过 sessionManager 触发
            sessionManager.triggerDuress()
        }

        // 擦除本地数据库
        try {
            val dbHelper = DatabaseHelper(context)
            dbHelper.wipeAllData()
            Log.i(TAG, "Local data wiped")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to wipe data: ${e.message}")
        }

        // 清除 SharedPreferences
        context.getSharedPreferences("spider_settings", Context.MODE_PRIVATE).edit().clear().apply()

        onDuressTriggered?.invoke()
    }

    /**
     * 设置胁迫 PIN。
     */
    fun setDuressPin(pin: String): Boolean {
        if (pin.length != 6 || !pin.matches(Regex("\\d{6}"))) return false
        keyManager.setDuressPin(pin)
        return true
    }

    fun hasDuressPin(): Boolean = keyManager.hasDuressPin()
}
